### ISSUE TITLE
GCS Soft Delete Announcement / Terraform interaction

### ISSUE THREAD

**Issue by joe-a-t** [hashicorp/terraform-provider-google#17127]
I just saw https://cloud.google.com/resources/storage/soft-delete-announce?hl=en and this raised a lot of questions about how Terraform will interact with deleting both objects and buckets once soft delete is enabled by default by GCP. Can provider maintainers please weigh in on how you plan to interact with GCS once this goes into effect? 

1. Will Terraform succeed in deleting an object/bucket if it is in soft delete state or will the object/bucket continuing to exist in that soft delete state instead of having been entirely deleted cause Terraform to fail?
2. How will Terraform respond when it goes to delete a bucket if all objects in the bucket are in the soft delete state? 
3. Will `force_destroy` continue to work? Looks like no based on my understanding.
4. Will the Terraform Google Provider set any default retention period settings that contradict the new default retention settings Google will be setting as part of this announcement? 
5. If the Terraform Google Provider does not set a default retention period, will the provider be modified to gracefully accept the new default value Google is setting?
6. etc

Feel free to modify this at will or replace with your own issue/post covering these issues, just trying to throw these out a starting point.

b/326044671

**Comment by shuyama1** [comment]
Thanks @joe-a-t for taking time to submit your questions and for opening this ticket. Seeing as this feature has not been launched officially, I will answer the questions I can with the knowledge I have now, and I will also be forwarding this ticket to the service team so they can provide their input.

> 1. Will Terraform succeed in deleting an object/bucket if it is in soft delete state or will the object/bucket continuing to exist in that soft delete state instead of having been entirely deleted cause Terraform to fail?

With the soft-delete feature being launched for GCS, I do not think we anticipate any significant changes to how Terraform handles deletion of these resources. Terraform should still be able to delete an object or bucket with a soft-delete policy enabled (which is the default setting). This means that running `terraform destroy` will not result in an error with soft delete enabled and the resource will simply enter the soft-delete state. One possible change is that a new resource may not be created if it has the same name as an existing resource that is still in the soft delete retention period.

> 2. How will Terraform respond when it goes to delete a bucket if all objects in the bucket are in the soft delete state?

I think it depends heavily on how the API handles this scenario. Terraform will generate `DELETE` call to the API and pass back whatever response the API returns.

> 3. Will `force_destroy` continue to work? Looks like no based on my understanding.

This also depends heavily on the implementation details of the feature at the backend. Based on the information I have now, it seems that even with `force_destroy`, the resource will still be soft deleted unless the soft deletion is explicitly disabled.

> 4. Will the Terraform Google Provider set any default retention period settings that contradict the new default retention settings Google will be setting as part of this announcement? 

No, we do not plan to set a client-side default for the soft delete retention period.

>  5. If the Terraform Google Provider does not set a default retention period, will the provider be modified to gracefully accept the new default value Google is setting?

Once the feature has been rolled out, Terraform Google Provider will honor the service-side default for the retention period. Soft-delete period related fields will need to be implemented in the providers to support customizing soft delete settings via Terraform after the API supports it.

**Comment by joe-a-t** [comment]
I think there is some confusion here? As far as I understand it, all of the relevant service side APIs exist today and you can test the exact performance of the proposed change today by explicitly setting a retention policy of 7 days on a [bucket via a retention_policy](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket#retention_policy). The contents of the [announcement](https://cloud.google.com/resources/storage/soft-delete-announce?hl=en) is not as much a new feature, as a change of the default value.

**Comment by shuyama1** [comment]
My understanding is that the Soft Delete feature is different from the current `retention_policy`. 

The current one is to ensure that the objects cannot be deleted at all until they reach the retention period specified. Before that, attempts to delete the objects will result in a failed deletion with the `403 - retentionPolicyNotMet` error, according to [Bucket retention policies](https://cloud.google.com/storage/docs/bucket-lock#retention-policy). 

The soft delete feature is, if enabled, after the object is deleted successfully, it goes into the soft-deleted state ("Soft-deleted objects are special non-readable objects that are hidden from [object listings](https://cloud.google.com/storage/docs/json_api/v1/objects/list) unless a specific option is specified"), with the capability to be restored, according to the [soft delete announcement](https://cloud.google.com/resources/storage/soft-delete-announce?hl=en#functional-overview). The soft delete retention period means how long the object will be retained before it's permanently deleted.

**Comment by joe-a-t** [comment]
Yeah, I think the normal retention policy is only for the "LIVE" version of an object since it's not compatible with object versioning. This new soft delete thing _seems_ like it is basically a retention policy on "ARCHIVED" versions? It's definitely weird that it is disabled via reusing the existing retention policy arguments if it behaves differently than retention policy currently does. 

So I guess it is super unclear if these _soft deleted objects_ will behave in the same way as "ARCHIVED" versions of objects when it comes to using Terraform to delete a bucket with `force_destroy = false`, which currently fails with 
```
Error: Error trying to delete bucket test containing objects without `force_destroy` set to true
```
if you try to delete a bucket that does not have any "LIVE" objects but does have some "ARCHIVED" objects. And there are a couple of possible ways I could see this being handled like:

1. Being unable to delete the bucket regardless of method until all objects have been deleted and the soft delete period on all objects has expired (most similar to current behavior if trying to delete a bucket that contains objects where the retention period is not met but appears to not be the case due to the last part of [How restores work](https://cloud.google.com/resources/storage/soft-delete-announce?hl=en#functional-overview)?)
2. Being able to delete the bucket via Terraform but only if you use `force_destroy = true` when there are not any "LIVE" objects but are still objects in the soft delete period.
3. Being able to delete the bucket via Terraform with `force_destroy = false` when there are still soft deleted objects in the bucket but no "LIVE" objects.

Of these options, #3 would probably be my personal preference and the least impactful to current Provider behavior although depending on how exactly it was implemented on the Google API side, there Provider may need to be modified so that it responds with a "yes I'm sure" upon receiving a warning that the bucket still contains soft deleted objects but no "LIVE" objects. 

And it sounds like for any information on how this would function, we are stuck waiting on the Google service team to provide feedback on how their changes will affect the relevant APIs and if those API changes result in any kinds of changes on the Terraform Provider side.

**Comment by shuyama1** [comment]
Changed the label to `bug`, as I just realize that questions are not properly forwarded internally.

**Note to the service team**: This particular ticket is not a bug but rather a question. We're classifying it as such to ensure proper forwarding, as our internal system does not have a corresponding ticket type. We’d like to invite the service team to provide insight on the upcoming soft deletion feature in GCS and its impact on current Terraform users, specifically whether it will affect the current Terraform deletion process. In Terraform Google Provider,  [bucket deletion method] (https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/services/storage/resource_storage_bucket.go#L868) offers an option `force_destroy` to whether to force delete a bucket by deleting all contained objects first, and then makes direct API calls to delete the bucket.

To better understand the potential impact for bucket deletion in Terraform, the more specific questions are:
1. Does `LIST` also show soft-deleted objects? It seems no based on the [announcement](https://cloud.google.com/resources/storage/soft-delete-announce?hl=en) 
2. Does `DELETE` succeed for deleting a bucket containing only soft-deleted objects? If no, this will break the current Terraform deletion.

Additionally, it seems that soft deletion will be enabled by default for both existing and new buckets/objects. Please be aware that this means users won't be able to turn off soft-deletion via Terraform until the feature is implemented into the terraform providers as well. 
Many thanks!

**Comment by kautikdk** [comment]
Hi @joe-a-t , @shuyama1. In response to above discussion thread, I like to highlight some technical aspects of upcoming soft delete feature.

1. Soft delete is not introducing any breaking changes to Cloud Storage. 
2. Buckets that contain soft-deleted objects can still be deleted, just as before. It is different than retention policy which can prevent bucket deletion in some cases.
3. The list API will not return soft-deleted objects unless the new softDeleted query parameter is specified.
4. Any settings you provide in a Terraform will override service defaults.

Let me know if it answer the questions or something is missing!

**Comment by joe-a-t** [comment]
@kautikdk Thank you. Can you please get this information clarified on https://cloud.google.com/resources/storage/soft-delete-announce?hl=en? 

In particular, 
![image](https://github.com/hashicorp/terraform-provider-google/assets/67290882/f18e3a8c-cb47-421d-9ac3-53386131fe8c)
was quite confusing in how it appears to now be talking about a separate soft-delete retention period, not the [existing bucket retention period](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket#retention_period). Given what appears to be huge differences in functionality, I'm quite worried about having identical vars that have very different purposes (and cost implications) depending on what specific block they are in. Might be better to change the name (maybe switch `retention period` to `duration`) or some other clarification.

Additionally, it would be helpful if the `Terraform` link at the bottom of the announcement linked to an open issue or PR on the Terraform provider that had further details about how it will be implemented or links to the docs/implementation on the google-beta provider instead of the generic Google page for "What is Terraform"

**Comment by kautikdk** [comment]
@joe-a-t Sure, These information will be clarified on documentation. More technical details will be available at launch time which is in near future.

In the above documentation it refers to soft delete retention period. Generally Soft delete terraform support will be implemented as new block as new terraform fields should mirror the API’s structure to ease predictability and maintenance. Inside that block there will be one field which can be used to specify soft delete duration in seconds. I will redirect your feedback regarding naming to appropriate team.

Regarding `Terraform` link, It is linked for tool reference rather than feature support reference. I am not sure about providing any link to open issue/PR/docs on terraform provider because providing terraform support for feature might have different priority considering CLI and Cloud Console support.

Thank you for your valuable feedback and views!

**Comment by NissesSenap** [comment]
So in the announcement, March was mentioned as the launch.
I can see soft delete in the UI, and it links to a document https://cloud.google.com/storage/docs/soft-delete, that shows 404.

![image](https://github.com/hashicorp/terraform-provider-google/assets/14774899/6ea2ae83-5144-4572-b588-c79e9cad118a)

@kautikdk any status update on this feature, and when this feature is coming to terraform would be great.
I have multiple places where this will need to be disabled, and currently GCP isn't giving us much time to do it.

**Comment by kautikdk** [comment]
Hi @NissesSenap, Terraform support will be provided after full GA of the feature but It will be before promotional time ends mentioned in the announcement which is May 31, 2024. 
Regarding the link, I will redirect it to the respective team working on this.

**Comment by NissesSenap** [comment]
To be honest, @kautikdk , that don't sit too good with me and I will reach out to our sales contacts to discuss this.

GCP are doing a big change in its API and you are giving us very limited timeframe to adjust to that.
If you at least can add this to the beta provider we will at least have more time to do some tests around this.
Since I see it in the UI, I assume that the API is already there.

In short, the quicker this can be implemented so we as your customers can start to try this the better. And we would be very grateful if this can be solved as quickly as possible.
